### PR TITLE
Makes possible to run all the tests and doctests under coverage

### PR DESCRIPTION
**Pull request by jcemelanda**
I want to try to create more tests as I have time

### CODE REVIEW

**Comment by gyermolenko** [comment]
@jcemelanda thanks! 
Please review existing tests on the same topic also.
And as an example for factory - here is a link for related doctest
https://github.com/faif/python-patterns/blob/master/patterns/creational/factory.py#L63

Also I think that doctests in general are sufficient and suit the idea better as they are more illustrative. But that could not be the case for some specific checks you have in mind.

**Comment by faif** [comment]
@jcemelanda Did you commit an empty file? I cannot view any changes/file contents

**Comment by jcemelanda** [comment]
> @jcemelanda Did you commit an empty file? I cannot view any changes/file contents

That's weird, I'll check it here but the commit had data

**Comment by jcemelanda** [comment]
> @jcemelanda thanks!
> Please review existing tests on the same topic also.
> And as an example for factory - here is a link for related doctest
> https://github.com/faif/python-patterns/blob/master/patterns/creational/factory.py#L63
> 
> Also I think that doctests in general are sufficient and suit the idea better as they are more illustrative. But that could not be the case for some specific checks you have in mind.

I like the doctests but they were not run when I tried to get the coverage....

**Comment by jcemelanda** [comment]
> @jcemelanda thanks!
> Please review existing tests on the same topic also.
> And as an example for factory - here is a link for related doctest
> https://github.com/faif/python-patterns/blob/master/patterns/creational/factory.py#L63
> 
> Also I think that doctests in general are sufficient and suit the idea better as they are more illustrative. But that could not be the case for some specific checks you have in mind.

BTW, maybe I missunderstood the use of tests in this project. Are they used only for usage test? Or are they actually intended to test? The run_all script is not running all the tests by what I get in the coverage report. That was my main reason to start writing this test.

I just don't know why it's missing here.

**Comment by gyermolenko** [comment]
@jcemelanda please ignore `run_all.sh`. AFAIK it was intended to run all .py files to get coverage.
At the moment coverage is checked by `pytest-cov` lib.

Please check your commit and make sure it has the changes you wanted to contribute.

**Comment by jcemelanda** [comment]
> @jcemelanda please ignore `run_all.sh`. AFAIK it was intended to run all .py files to get coverage.
> At the moment coverage is checked by `pytest-cov` lib.
> Easiest way to check what it does locally is:
> 
> ```
> $ pip install -r requirements-dev.txt
> $ pytest --cov=patterns tests/
> 
> Name                                                   Stmts   Miss  Cover
> --------------------------------------------------------------------------
> ...
> patterns/creational/factory.py                            18     18     0%
> ...
> --------------------------------------------------------------------------
> TOTAL                                                   1496   1005    33%
> ```
> 
> And "0%" is misleading here because it doesn't include doctests. I want to fix this at some point.
> 
> Please check your commit and make sure it has the changes you wanted to contribute.

Great! I'll finish it today.

Thanks for the info.

**Comment by jcemelanda** [comment]
> @jcemelanda please ignore `run_all.sh`. AFAIK it was intended to run all .py files to get coverage.
> At the moment coverage is checked by `pytest-cov` lib.
> Easiest way to check what it does locally is:
> 
> ```
> $ pip install -r requirements-dev.txt
> $ pytest --cov=patterns tests/
> 
> Name                                                   Stmts   Miss  Cover
> --------------------------------------------------------------------------
> ...
> patterns/creational/factory.py                            18     18     0%
> ...
> --------------------------------------------------------------------------
> TOTAL                                                   1496   1005    33%
> ```
> 
> And "0%" is misleading here because it doesn't include doctests. I want to fix this at some point.
> 
> Please check your commit and make sure it has the changes you wanted to contribute.

Based on your comment I relized it'd be better to just make the run_all.sh run all the tests with the doctests on the coverage.

I deleted the new test file I created.

File: run_all.sh [hunk 1]
```diff
@@ -6,17 +6,4 @@
...
-    PYTHONPATH=. python $COVERAGE $f || failed+=" $f"
-    echo "I: done $f. Exit code $?"
-done;
-
-if [ ! -z "$failed" ]; then
-    echo "Failed: $failed";
-    exit 1
-fi
\ No newline at end of file
+pytest --doctest-modules --cov=patterns
```
[the remark this answers is not in the document]
Comment (jcemelanda) [reply run_all.sh]: Ok, so we should remove the file if it should not be used.

BTW when I run only pytest --doctest-modules --cov=patterns
It does run the doctests

> patterns/creational/builder.py                                  39      8    79%
> patterns/creational/factory.py                                  18      2    89%
> patterns/creational/lazy_evaluation.py                    35      3    91%

[the remark this answers is not in the document]
Comment (gyermolenko) [reply run_all.sh]: You are right, I remembered it incorrectly.
```
$ pytest --cov=patterns | grep "\bfactory.py"
patterns/creational/factory.py                            18     18     0%

$ pytest --cov=patterns --doctest-modules | grep "\bfactory.py"
patterns/creational/factory.py .                                         [ 17%]
patterns/creational/factory.py                            18      2    89%
```

` --doctest-modules` does help with coverage report

You can investigate it further with commands from tox.ini.

**Comment by faif** [comment]
@jcemelanda Can you remove commit ec275a9f0f96cfaa34d0b7143bf077d1f162273e from this pull request? It doesn't seem to do anything useful :)

**Comment by faif** [comment]
Please resolve the conflicts

**Comment by jcemelanda** [comment]
Hi! I was reviewing my code and the comments here and I think I misunderstood how the tests were run. I think I should close this PR and come back to it another time.

**Pull request closed without merging**